general/kdtree.hpp

```diff
@@ -161,34 +182,25 @@ class KDTree
 
    /// Builds the KDTree. If the point cloud is modified the tree
    /// needs to be rebuild by a new call to Sort().
-   void Sort()
+   virtual void Sort() override
```

Comment:
These are virtual because they are part of the `KDTreeBase` interface introduced in this PR. (However, there is no need to have them listed as `virtual` in the derived class, just `override` is enough — is that what you are asking about?)

Reply:
Yes, I was referring to the derived class. Perhaps there is already a use case for another child class in a future development.

Reply:
`virtual` should not be used to define methods in child classes, as it does not explicitly bind the relationship to the parent method. For instance were the parent class method name to change (or be deleted), I believe this would still compile, and the two methods would now decouple. Using `override` is strictly better in child classes as it expresses the intent in language that the compiler can understand, allowing it to strictly enforce the interface. As a silly example, with just virtual, the parent method and child method could have different `const` correctness and coexist, then the dispatch would be purely on the `const` status rather than vtable. With override this becomes impossible as the compiler errors that the `override` method signature does not match the parent. I've caught a few bugs in various classes in mfem just by making this change in derived classes as I come across them.

If you want to prevent there being further children redefining the method, then `final` is the identifier to use, but that only has value in tightly constraining interfaces or for devirtualization, which are both rarely required.

Reply:
In this case, if the base class were to be changed, it would not compile, since the member functions in the derived class are marked `override`. (But as mentioned above, the `virtual` doesn't add anything, so it can be removed—done in 7c9abd530).